File: app/src/main/java/com/karsu/ballonsmenu/ListenerActivity.java
/** Created by Erkan Kaplan on 2026-02-03 */
package com.karsu.ballonsmenu;

import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import android.util.Log;
import android.widget.TextView;

import com.karsu.ballonsmenu.BoomButtons.BoomButton;
import com.karsu.ballonsmenu.BoomButtons.ButtonPlaceEnum;
import com.karsu.ballonsmenu.BoomButtons.OnBMClickListener;
import com.karsu.ballonsmenu.BoomButtons.SimpleCircleButton;
import com.karsu.ballonsmenu.BoomMenuButton;
import com.karsu.ballonsmenu.ButtonEnum;
import com.karsu.ballonsmenu.OnBoomListener;
import com.karsu.ballonsmenu.OnBoomListenerAdapter;
import com.karsu.ballonsmenu.Piece.PiecePlaceEnum;

public class ListenerActivity extends AppCompatActivity {

    private TextView textViewForAnimation;
    private TextView textViewForButton;
    private BoomMenuButton bmb;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_listener);

        textViewForButton = (TextView) findViewById(R.id.text_for_button);
        textViewForAnimation = (TextView) findViewById(R.id.text_for_animation);

        bmb = (BoomMenuButton) findViewById(R.id.bmb);
        assert bmb != null;
        bmb.setButtonEnum(ButtonEnum.SimpleCircle);
        bmb.setPiecePlaceEnum(PiecePlaceEnum.DOT_6_3);
        bmb.setButtonPlaceEnum(ButtonPlaceEnum.SC_6_3);
        for (int i = 0; i < bmb.getPiecePlaceEnum().pieceNumber(); i++) addBuilder();

        // Use OnBoomListenerAdapter to listen part of methods
        bmb.setOnBoomListener(new OnBoomListenerAdapter() {
            @Override
            public void onBoomWillShow() {
                super.onBoomWillShow();
                // logic here
            }
        });

        // Use OnBoomListener to listen all methods
        bmb.setOnBoomListener(new OnBoomListener() {
            @Override
            public void onClicked(int index, BoomButton boomButton) {
                // If you have implement listeners for boom-buttons in builders,
                // then you shouldn't add any listener here for duplicate callbacks.
            }

            @Override
            public void onBackgroundClick() {
                textViewForAnimation.setText("Click background!!!");
            }

            @Override
            public void onBoomWillHide() {
                Log.d("BMB", "onBoomWillHide: " + bmb.isBoomed() + " " + bmb.isReBoomed());
                textViewForAnimation.setText("Will RE-BOOM!!!");
            }

            @Override
            public void onBoomDidHide() {
                Log.d("BMB", "onBoomDidHide: " + bmb.isBoomed() + " " + bmb.isReBoomed());
                textViewForAnimation.setText("Did RE-BOOM!!!");
            }

            @Override
            public void onBoomWillShow() {
                Log.d("BMB", "onBoomWillShow: " + bmb.isBoomed() + " " + bmb.isReBoomed());
                textViewForAnimation.setText("Will BOOM!!!");
            }

            @Override
            public void onBoomDidShow() {
                Log.d("BMB", "onBoomDidShow: " + bmb.isBoomed() + " " + bmb.isReBoomed());
                textViewForAnimation.setText("Did BOOM!!!");
            }
        });
    }

    private void addBuilder() {
        bmb.addBuilder(new SimpleCircleButton.Builder()
                .normalImageRes(BuilderManager.getImageResource())
                .listener(new OnBMClickListener() {
                    @Override
                    public void onBoomButtonClick(int index) {
                        textViewForButton.setText("No." + index + " boom-button is clicked!");
                    }
                }));
    }
}

File: app/src/main/kotlin/com/karsu/ballonsmenu/ListenerActivity.kt
/** Created by Erkan Kaplan on 2026-02-03 */
package com.karsu.ballonsmenu

import android.os.Bundle
import android.util.Log
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.karsu.ballonsmenu.app.R
import com.karsu.ballonsmenu.KarSuButtons.KarSuButton
import com.karsu.ballonsmenu.KarSuButtons.ButtonPlaceEnum
import com.karsu.ballonsmenu.KarSuButtons.SimpleCircleButton
import com.karsu.ballonsmenu.KarSuMenuButton
import com.karsu.ballonsmenu.ButtonEnum
import com.karsu.ballonsmenu.OnKarSuListener
import com.karsu.ballonsmenu.Piece.PiecePlaceEnum

class ListenerActivity : AppCompatActivity() {

    private lateinit var textViewForAnimation: TextView
    private lateinit var textViewForButton: TextView
    private lateinit var bmb: KarSuMenuButton

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_listener)

        textViewForButton = findViewById(R.id.text_for_button)
        textViewForAnimation = findViewById(R.id.text_for_animation)

        bmb = findViewById<KarSuMenuButton>(R.id.bmb).apply {
            buttonEnum = ButtonEnum.SimpleCircle
            piecePlaceEnum = PiecePlaceEnum.DOT_6_3
            buttonPlaceEnum = ButtonPlaceEnum.SC_6_3
            repeat(piecePlaceEnum.pieceNumber()) {
                addBuilder()
            }
        }

        // Use OnKarSuListener to listen all methods
        bmb.onKarSuListener = object : OnKarSuListener {
            override fun onKarSuButtonClick(index: Int) {
                // If you have implement listeners for boom-buttons in builders,
                // then you shouldn't add any listener here for duplicate callbacks.
            }

            override fun onBackgroundClick() {
                textViewForAnimation.text = "Click background!!!"
            }

            override fun onKarSuWillHide() {
                Log.d("BMB", "onKarSuWillHide: ${bmb.isKarSued()} ${bmb.isReKarSued()}")
                textViewForAnimation.text = "Will RE-BOOM!!!"
            }

            override fun onKarSuDidHide() {
                Log.d("BMB", "onKarSuDidHide: ${bmb.isKarSued()} ${bmb.isReKarSued()}")
                textViewForAnimation.text = "Did RE-BOOM!!!"
            }

            override fun onKarSuWillShow() {
                Log.d("BMB", "onKarSuWillShow: ${bmb.isKarSued()} ${bmb.isReKarSued()}")
                textViewForAnimation.text = "Will BOOM!!!"
            }

            override fun onKarSuDidShow() {
                Log.d("BMB", "onKarSuDidShow: ${bmb.isKarSued()} ${bmb.isReKarSued()}")
                textViewForAnimation.text = "Did BOOM!!!"
            }
        }
    }

    private fun addBuilder() {
        bmb.addBuilder(
            SimpleCircleButton.Builder()
                .normalImageRes(BuilderManager.getImageResource())
                .listener { index ->
                    textViewForButton.text = "No.$index boom-button is clicked!"
                }
        )
    }
}
